lib/plugin/registry.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2015-2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugin
+
+import "github.com/gravitational/trace"
+
+// Plugin describes interfaces of the teleport core plugin
+type Plugin interface {
+	// GetName returns plugin name
+	GetName() string
+	// RegisterProxyWebHandlers registers new methods with the ProxyWebHandler
+	RegisterProxyWebHandlers(handler interface{}) error
+	// RegisterAuthWebHandlers registers new methods with the Auth Web Handler
+	RegisterAuthWebHandlers(service interface{}) error
+	// RegisterAuthServices registers new services on the AuthServer
+	RegisterAuthServices(server interface{}) error
+}
+
+// Registry is the plugin registry
+type Registry interface {
+	// Add adds plugin to the registry
+	Add(plugin Plugin) error
+	// RegisterProxyWebHandlers registers Teleport Proxy web handlers
+	RegisterProxyWebHandlers(hander interface{}) error
+	// RegisterAuthWebHandlers registers Teleport Auth web handlers
+	RegisterAuthWebHandlers(handler interface{}) error
+	// RegisterAuthServices registerse Teleport AuthServer services
+	RegisterAuthServices(server interface{}) error
+}
+
+// NewRegistry creates an instance of the Registry
+func NewRegistry() Registry {
+	return &registry{
+		plugins: make(map[string]Plugin),
+	}
+}
+
+type registry struct {
+	plugins map[string]Plugin
```

Comment:
This map is not going to be used concurrently as plugin registrations happen during the initialization time. 

Btw, @klizhentas i am not happy with the package name `plugin` as it overlaps with other plugins (data). What if I name it as `enterprise` implying enterprise only extensions?

Reply:
I think either `plugin` or `extensions` is fine. I would stay away from `enterprise`